packages/apollo-server-core/src/requestPipelineAPI.ts

```diff
@@ -10,3 +23,145 @@ export {
   GraphQLExecutor,
   GraphQLExecutionResult,
 } from 'apollo-server-types';
+
+export const symbolRequestListenerDispatcher =
+  Symbol("apolloServerRequestListenerDispatcher");
+export const symbolPluginsEnabled = Symbol("apolloServerPluginsEnabled");
+
+export function enablePluginsForSchemaResolvers(
+  schema: GraphQLSchema & { [symbolPluginsEnabled]?: boolean },
+) {
+  if (schema[symbolPluginsEnabled]) {
+    return schema;
+  }
+  Object.defineProperty(schema, symbolPluginsEnabled, {
+    value: true,
+  });
+
+  forEachField(schema, wrapField);
+
+  return schema;
+}
+
+function wrapField(field: GraphQLField<any, any>): void {
+  const fieldResolver = field.resolve || defaultFieldResolver;
+
+  field.resolve = (source, args, context, info) => {
+    // This is a bit of a hack, but since `ResponsePath` is a linked list,
+    // a new object gets created every time a path segment is added.
+    // So we can use that to share our `whenObjectResolved` promise across
+    // all field resolvers for the same object.
+    const parentPath = info.path.prev as ResponsePath & {
+      __fields?: Record<string, ReadonlyArray<FieldNode>>;
+      __whenObjectResolved?: Promise<any>;
+    };
+
+    // The technique for implementing a  "did resolve field" is accomplished by
+    // returning a function from the `willResolveField` handler.  The
+    // dispatcher will return a callback which will invoke all of those handlers
+    // and we'll save that to call when the object resolution is complete.
+    const endHandler = context && context[symbolRequestListenerDispatcher] &&
+      (context[symbolRequestListenerDispatcher] as Dispatcher<GraphQLRequestListener>)
+        .invokeDidStartHook('willResolveField', source, args, context, info) ||
+          ((_err: Error | null, _result?: any) => { /* do nothing */ });
+
+    const resolveObject: GraphQLObjectResolver<
+      any,
+      any
+    > = (info.parentType as any).resolveObject;
+
+    let whenObjectResolved: Promise<any> | undefined;
+
+    if (parentPath && resolveObject) {
+      if (!parentPath.__fields) {
+        parentPath.__fields = {};
+      }
+
+      parentPath.__fields[info.fieldName] = info.fieldNodes;
+
+      whenObjectResolved = parentPath.__whenObjectResolved;
+      if (!whenObjectResolved) {
+        // Use `Promise.resolve().then()` to delay executing
+        // `resolveObject()` so we can collect all the fields first.
+        whenObjectResolved = Promise.resolve().then(() => {
+          return resolveObject(source, parentPath.__fields!, context, info);
+        });
```

Comment:
I don't actually understand what this comment is trying to explain. Where are we collecting fields (presumably asynchronously) that we need to defer L87?

Reply:
Worth noting that this is unchanged from the `graphql-extensions` implementation which this is lifted from; I take no credit for this, but it's been working for a while.  I suspect this is rigged this way to wrap each of the fields on this type — and allow the plugin's implementation to do what it would like to those — before initiating the resolvers for those fields (and another recursive invocation of the wrapped field resolver in which this code lives).